### PR TITLE
Eliminate ConstnessAnd again

### LINKED ISSUE #89432: nightly-2021-10-01: the trait `From<&str>` is not implemented for `std::string::String`
I tried this code:

```rust
pub struct Datum {
    /// The name of this datum.
    pub name: String,

    /// Whether the datum indicates support for the platform or not.
    pub pass: bool,

    /// Short additional information to display to the user.
    pub info: Option<String>,

    /// Longer explanatory message on how to resolve problems.
    pub mesg: Option<String>,
}
```

```rust
fn dev_kvm() -> super::Datum {
    let dev_kvm = std::path::Path::new("/dev/kvm");

    super::Datum {
        name: "Driver".into(),
        pass: dev_kvm.exists(),
        info: Some("/dev/kvm".into()),
        mesg: None,
    }
}
```

I expected to see this happen:

Successful compilation. This has compiled and worked for years until `nightly-2021-10-01`. Reverting to `nightly-2021-09-30` produces a successful compilation.

Instead, this happened:

```
error[E0277]: the trait bound `std::string::String: From<&str>` is not satisfied
  --> src/backend/kvm/mod.rs:23:24
   |
23 |         name: "Driver".into(),
   |                        ^^^^ the trait `From<&str>` is not implemented for `std::string::String`
   |
   = note: required because of the requirements on the impl of `Into<std::string::String>` for `&str`
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
   |
19 | fn dev_kvm() -> super::Datum where std::string::String: From<&str> {
   |                              +++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `std::string::String: From<&str>` is not satisfied
  --> src/backend/kvm/mod.rs:25:20
   |
25 |         info: Some("/dev/kvm".into()),
   |               ---- ^^^^^^^^^^^^^^^^^ the trait `From<&str>` is not implemented for `std::string::String`
   |               |
   |               required by a bound introduced by this call
   |
   = note: required because of the requirements on the impl of `Into<std::string::String>` for `&str`
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
   |
19 | fn dev_kvm() -> super::Datum where std::string::String: From<&str> {
   |                              +++++++++++++++++++++++++++++++++++++
```

### Meta

`rustc --version --verbose`:
```
rustc 1.57.0-nightly (aa7aca3b9 2021-09-30)
binary: rustc
commit-hash: aa7aca3b954131720df725e70d12e902eb3be1de
commit-date: 2021-09-30
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

cc @haraldh @wgwoods

### LINKED ISSUE #91489: nightly-2021-12-02 to nightly-2021-12-03 regression (PR #91354): auto-deref on `Cow` does not seem to work anymore in all circumstances
### Code

See https://github.com/gtk-rs/gtk-rs-core/blob/master/glib-macros/tests/test.rs#L290

This fails with the latest nightly with (run `cargo test` inside the `glib-macros` directory).

```
error[E0599]: no method named `as_str` found for enum `Cow<'static, VariantTy>` in the current scope
   --> glib-macros/tests/test.rs:290:48
    |
290 |     assert_eq!(Variant1::static_variant_type().as_str(), "(si)");
    |                         
```

Simplified code from `gtk-rs` / `glib` which works without problems for some reason so I'm not entirely sure what causes it:

<details>

```rust
pub struct VariantType;
pub struct VariantTy;

impl Borrow<VariantTy> for VariantType {
    fn borrow(&self) -> &VariantTy {
        todo!()
    }
}

impl ToOwned for VariantTy {
    type Owned = VariantType;

    fn to_owned(&self) -> VariantType {
        todo!()
    }
}

impl VariantTy {
    pub fn as_str(&self) -> &str {
        todo!()
    }
}

let x = Cow::Owned(VariantType);
println!("{}", x.as_str());
```

</details>

### Version it worked on

It most recently worked on: 2021-12-02

### Version with regression

`rustc --version --verbose`:

```
rustc 1.59.0-nightly (acbe4443c 2021-12-02)
binary: rustc
commit-hash: acbe4443cc4c9695c0b74a7b64b60333c990a400
commit-date: 2021-12-02
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```

(note that this is actually the one from today, 2021-12-03, but there's an off-by-one in the version information)

Running `cargo bisect-rustc` yields:

```
searched nightlies: from nightly-2021-12-01 to nightly-2021-12-03
regressed nightly: nightly-2021-12-03
searched commits: from https://github.com/rust-lang/rust/commit/48a5999fceeea84a8971634355287faa349909d4 to https://github.com/rust-lang/rust/commit/acbe4443cc4c9695c0b74a7b64b60333c990a400
regressed commit: https://github.com/rust-lang/rust/commit/18bb8c61a975fff6424cda831ace5b0404277145
```

Which would be this PR: https://github.com/rust-lang/rust/pull/91354 (CC @fee1-dead).

@rustbot modify labels: +regression-from-stable-to-nightly -regression-untriaged

### PR DESCRIPTION
**Pull request by fee1-dead**
Closes #91489.
Closes #89432.

Reverts #91491.
Reverts #89450.

r? @spastorino

### CODE REVIEW

**Comment by oli-obk** [comment]
@bors r-

Should check whether this also solves the duplicate query evals

**Comment by fee1-dead** [comment]
Requesting another perf run as I have tried to improve performance in recent commits.

@bors try @rust-timer queue

**Comment by fee1-dead** [comment]
@rustc-timer build a4303f77cdf54b605cd5eb5b74c04d0611695f1a

**Comment by fee1-dead** [comment]
I'm seeing some regressions in `evaluate_obligations` query. The `eval_to_allocation_raw` is probably spurious related to CGU division and PGO. Pushing a fix in a moment for `evaluate_obligations`.

**Comment by fee1-dead** [comment]
I don't really have an idea for the reason of this regression. If I have time I am going to benchmark this locally and see what callgrind says about this change.

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 1]
```diff
@@ -918,6 +918,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         } else {
             self.param_env
         };
+        let param_env = param_env.with_const();
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:921]: From past experience with similar regressions I would guess that this causes inlining of `eval_to_allocation` to change, messing up other LLVM optimizations at the caller sites or at other function calls within in this function.

Comment (fee1-dead) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:921]: Should I add `#[inline]` to these functions or should I remove these calls and assertions altogether?

Comment (oli-obk) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:921]: Either profile to see what is now not inlined, or add inline (maybe even always or never) to this function, or any of the functions it calls. I'm not sure what exactly is needed, and only profiling can tell for sure

**Comment by fee1-dead** [comment]
I got the ctfe regression down to ~0.9% with the latest change based on my local benchmarks. Upon further investigation I don't think there is an easy way to fix the remaining regressions. I found the `mir::Body::dominators` not being inlined in `mir_borrowck` with this PR which causes it to regress.

**Comment by oli-obk** [comment]
I think we can live with that regression, there are so many similar inlining perf losses lying around that we should address properly, this PR should not be blocked on it. It makes the logic saner after all

**Comment by fee1-dead** [comment]
Anyways, let's check one last time (in case my local measurements are inaccurate):

@bors try @rust-timer

**Comment by oli-obk** [comment]
@rustbot label: +perf-regression-triaged

This is a significant simplification of various ad-hoc schemes that were introduced for const trait bounds. It now does this correctly by tracking it in the parameter environment and in the `TraitPredicate` instead of re-creating these datastructures on the fly with tuples.

The regression in const eval is likely due how LLVM inlines things, as we've seen the same regression happen in other PRs that touched this region of code in the CTFE machinery. The game of whack-a-mole with that regression is not fun, and we should just merge the PR and do a proper analysis and overhaul of the CTFE machinery and see where we can improve perf.

**Pull request merged**